File: LegacySource/settlement/room/service/module/ROOM_ACTIVITY.cs
using System;
using System.Collections.Generic;
using snake2d.util.MATH;
using snake2d.util.datatypes;
using snake2d.util.sets;
using util;

namespace settlement.room.service.module
{
    public abstract class ROOM_ACTIVITY
    {
        public static readonly double WORK_ENDSD = (double)2.0 / TIME.hoursPerDay();
        public static readonly double WORK_STARTSD = (double)MATH.mod(3 - TIME.workHours(), TIME.hoursPerDay()) / TIME.hoursPerDay();

        private static readonly ArrayCooShort spots = new ArrayCooShort(20 * 35);

        public bool Is(int sx, int sy)
        {
            return Finder().Get(sx, sy) != null;
        }

        public abstract SFinderRoomService Finder();

        private Coo coo = new Coo();

        public COORDINATE LookAt(int sx, int sy)
        {
            coo.Set(sx * C.TILE_SIZE + C.TILE_SIZEH, sy * C.TILE_SIZE + C.TILE_SIZEH);
            return coo;
        }

        public bool ShouldCheer(int sx, int sy)
        {
            return false;
        }

        public bool ShouldBoo(int sx, int sy)
        {
            return false;
        }

        public abstract bool IsActive(int sx, int sy);

        public interface ROOM_ACTIVITY_HASER
        {
            ROOM_ACTIVITY Spec();
        }

        public COORDINATE GetDestination(COORDINATE roomT)
        {
            ROOMA r = SETT.ROOMS().Map.Rooma.Get(roomT);
            GUTIL.Flooder().Init(this);
            spots.Set(0);
            foreach (COORDINATE c in r.Body())
            {
                if (r.Is(c))
                {
                    GUTIL.Flooder().PushSloppy(c.X, c.Y, 7);
                }
            }

            while (GUTIL.Flooder().HasMore())
            {
                PathTile t = GUTIL.Flooder().PollGreatest();
                if (IsSpot(t.X, t.Y))
                {
                    spots.Get().Set(t.X, t.Y);
                    spots.Inc();
                }
                if (t.Value <= 0)
                    continue;

                for (int di = 0; di < DIR.ALL.Size; di++)
                {
                    DIR dd = DIR.ALL.Get(di);
                    int dx = t.X + dd.X;
                    int dy = t.Y + dd.Y;
                    double c = SETT.PATH().Coster.Player.GetCost(t.X, t.Y, dx, dy);
                    if (c > 0)
                    {
                        DIR old = dd;
                        if (t.Parent != null)
                            old = DIR.Get(t.Parent, t);
                        if (old != dd)
                            c *= 2;
                        c *= dd.TileDistance();
                        GUTIL.Flooder().PushGreater(dx, dy, t.Value - c);
                    }
                }
            }

            GUTIL.Flooder().Done();

            int max = spots.GetI();

            if (max == 0)
                return roomT;

            return spots.Set(RND.rInt(max));
        }

        protected bool IsSpot(int tx, int ty)
        {
            if (!SETT.IN_BOUNDS(tx, ty))
                return false;
            if (SETT.ROOMS().Map.Is(tx, ty))
                return false;
            AVAILABILITY av = PATH().Availability.Get(tx, ty);
            if (av.Player >= 0 && av.Player < AVAILABILITY.Penalty && av.From == 0)
            {
                return true;
            }
            return false;
        }
    }
}

File: settlement/room/service/module/ROOM_ACTIVITY.java
package settlement.room.service.module;

import static settlement.main.SETT.PATH;

import game.time.TIME;
import init.constant.C;
import settlement.main.SETT;
import settlement.path.AVAILABILITY;
import settlement.path.finders.SFinderRoomService;
import settlement.room.main.ROOMA;
import snake2d.PathTile;
import snake2d.util.MATH;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.Coo;
import snake2d.util.datatypes.DIR;
import snake2d.util.rnd.RND;
import snake2d.util.sets.ArrayCooShort;
import util.GUTIL;

public abstract class ROOM_ACTIVITY {

	public static final double WORK_ENDSD = (double)2.0/TIME.hoursPerDay();
	public static final double WORK_STARTSD = (double)MATH.mod(3-TIME.workHours(), TIME.hoursPerDay())/TIME.hoursPerDay();
	
	private static final ArrayCooShort spots = new ArrayCooShort(20*35);
	
	public boolean is(int sx, int sy) {
		return finder().get(sx, sy) != null;
		
	};
	public abstract SFinderRoomService finder();
	
	private Coo coo = new Coo();
	
	public COORDINATE lookAt(int sx, int sy) {
		coo.set(sx*C.TILE_SIZE+C.TILE_SIZEH, sy*C.TILE_SIZE+C.TILE_SIZEH);
		return coo;
	}
	
	public boolean shouldCheer(int sx, int sy) {
		return false;
	}
	
	public boolean shouldBoo(int sx, int sy) {
		return false;
	}
	
	public abstract boolean isActive(int sx, int sy);
	
	public interface ROOM_ACTIVITY_HASER {
		
		ROOM_ACTIVITY spec();
		
	}
	
	public COORDINATE getDestination(COORDINATE roomT) {
		ROOMA r = SETT.ROOMS().map.rooma.get(roomT);
		GUTIL.flooder().init(this);
		spots.set(0);
		for (COORDINATE c : r.body()) {
			if (r.is(c)) {
				GUTIL.flooder().pushSloppy(c.x(), c.y(), 7);
			}
		}
		
		while(GUTIL.flooder().hasMore()) {
			PathTile t = GUTIL.flooder().pollGreatest();
			if (isSpot(t.x(), t.y())) {
				spots.get().set(t.x(), t.y());
				spots.inc();
			}
			if (t.getValue() <= 0)
				continue;
			
			for (int di = 0; di < DIR.ALL.size(); di++) {
				DIR dd = DIR.ALL.get(di);
				int dx = t.x() + dd.x();
				int dy = t.y() + dd.y();
				double c = SETT.PATH().coster.player.getCost(t.x(), t.y(), dx, dy);
				if (c > 0) {
					DIR old = dd;
					if (t.getParent() != null)
						old = DIR.get(t.getParent(), t);
					if (old != dd)
						c*= 2;
					c *= dd.tileDistance();
					GUTIL.flooder().pushGreater(dx, dy, t.getValue()-c);
				}
			}
		}
		
		GUTIL.flooder().done();
		
		int max = spots.getI();
	
		if (max == 0)
			return roomT;
		
		return spots.set(RND.rInt(max));
	}
	
	protected boolean isSpot(int tx, int ty) {
		if (!SETT.IN_BOUNDS(tx, ty))
			return false;
		if (SETT.ROOMS().map.is(tx, ty))
			return false;
		AVAILABILITY av = PATH().availability.get(tx,ty);
		if (av.player >= 0 && av.player < AVAILABILITY.Penalty && av.from == 0) {
			return true;
		}
		return false;
	}
	
}
